gpkit/nomials/map.py

```diff
@@ -0,0 +1,181 @@
+        """Applies substitutions to a NomialMap
+
+        Parameters
+        ----------
+        substitutions : (dict-like)
+            list of substitutions to perform
+
+        varkeys : (set-like)
+            varkeys that are present in self
+            (externally provided to promote caching)
+
+        parsedsubs : bool
+            flag if the substitutions have already been parsed
+            to contain only keys in varkeys
+
+        """
+        # pylint: disable=too-many-locals, too-many-branches
+        if parsedsubs or not substitutions:
+            fixed = substitutions
+        else:
+            fixed = parse_subs(varkeys, substitutions)
+
+        if not substitutions:
+            if not self.expmap:
+                self.expmap, self.csmap = {exp: exp for exp in self}, {}
+            return self
+
+        cp = NomialMap()
+        cp.units = self.units
+        cp.expmap, cp.csmap = {}, {}
+        varlocs = defaultdict(set)
+        for exp, c in self.items():
+            new_exp = HashVector(exp)
+            cp.expmap[exp] = new_exp
+            cp[new_exp] = c
+            for vk in new_exp:
+                varlocs[vk].add((exp, new_exp))
+
+        for vk in varlocs:
+            if vk in fixed:
+                expval = []
+                exps, cval = varlocs[vk], fixed[vk]
+                if isinstance(cval, Strings):
+                    descr = dict(vk.descr)
+                    del descr["name"]
+                    cval = VarKey(name=cval, **descr)
+                if hasattr(cval, "hmap"):
+                    expval, = cval.hmap.keys()  # TODO: catch "can't-sub-posys"
+                    cval = cval.hmap  # pylint: disable=redefined-variable-type
+                if hasattr(cval, "to"):
+                    cval = mag(cval.to(vk.units or DIMLESS_QUANTITY))
+                    if isinstance(cval, NomialMap) and cval.keys() == [{}]:
+                        cval, = cval.values()
+                if expval:
+                    cval, = cval.values()
+                exps_covered = set()
+                for o_exp, exp in exps:
+                    x = exp[vk]
+                    powval = float(cval)**x if cval != 0 or x >= 0 else np.inf
+                    cp.csmap[o_exp] = powval * cp.csmap.get(o_exp, self[o_exp])
+                    if exp in cp and exp not in exps_covered:
+                        c = cp.pop(exp)
+                        del exp[vk]
+                        for key in expval:
+                            exp[key] = expval[key]*x + exp.get(key, 0)
+                        exp._hashvalue = None  # reset hash, # pylint: disable=protected-access
+                        cp[exp] = powval * c + cp.get(exp, 0)
+                        exps_covered.add(exp)
+                        cp.remove_zeros(only_check_cs=True)  # pylint: disable=protected-access
+        return cp
+
+    def mmap(self, orig):
+        "Maps substituted monomials back to the original nomial"
```

Comment:
Still think this needs a more detailed explanation (need not be long) -- specifically, what are the types of `orig`, `pmap`, and `m_from_ms`, and for those that are dicts or dict like, state the types as `{type1: type2}`